### PR TITLE
⚠️ Machine ProviderID equality is now strictly enforced

### LINKED ISSUE #4526: ProviderID set by capi infra providers should match the one set by the controller manager cloud-provider
**What steps did you take and what happened:**
In the existing providerID logic assumptions, cloudProvider and ID is what is used to compare, skipping the other segments of the providerID.  See https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/noderefutil/providerid.go#L86

This makes the assumption - not necessarily true - that IDs in different regions/zones won't be reused by the cloud provider.

**What did you expect to happen:**
I'd be in favour of changing the expectation for the cluster-api-providers to set exactly the same providerID the controller manager cloud-provider sets.

We actually ensured this in AWS a while ago kubernetes-sigs/cluster-api-provider-aws#1693.
Though there might be other provider specific reasons why this is not possible I'm not aware of.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api version:
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

### PR DESCRIPTION
**Pull request by jackfrancis**
**What this PR does / why we need it**:

This PR updates the capi `ProviderID` type equality enforcement (via the `Equals` object pointer receiver method) to validate against the entire provider ID string rather than a concatenation of the "cloud provider" plus "ID" substrings.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4526

### CODE REVIEW

**Comment by jackfrancis** [comment]
@enxebre @alexeldeib is this PR getting close to addressing #4526? Obviously simply changing the capi equality implementation will require per-cloud-provider changes in their cluster-api providers in order to prevent breakage.

Is that what we want to advocate?

**Comment by alexeldeib** [comment]
yeah this solves it. I actually don't know that it requires infra provider changes -- CAPI should not have been the one setting these values ideally, CCM or similar would be, so this PR just aligns CAPI to CCM's behavior. 

put differently: in AWS one node may be represented with multiple providerIDs, but a concrete node will only ever have one form of that provider ID, so I feel this should be safe.

probably good to get some CAPA eyes on this and maybe something less cloud-y like CAPD?

**Comment by jackfrancis** [comment]
@sedefsavas can you PTAL at this change and confirm what @alexeldeib is suggesting

thank you!

@CecileRobertMichon are you able to speak authoritiatively on capd ramifications?

**Comment by CecileRobertMichon** [comment]
> maybe something less cloud-y like CAPD?

CAPD runs the PR e2e tests so if tests are passing I feel good about this change. Let's run the full suite to be sure.

/test ls

File: controllers/noderefutil/providerid.go [hunk 1]
```diff
@@ -89,7 +89,7 @@ func (p *ProviderID) ID() string {
 
 // Equals returns true if both the CloudProvider and ID match.
 func (p *ProviderID) Equals(o *ProviderID) bool {
```
Comment (killianmuldoon) [inline controllers/noderefutil/providerid.go:91]: Should this be deprecated and replaced instead of changing the behaviour?

**Comment by enxebre** [comment]
>yeah this solves it. I actually don't know that it requires infra provider changes -- CAPI should not have been the one setting these values ideally, CCM or similar would be, so this PR just aligns CAPI to CCM's behavior.

CCM set on Nodes but capi providers set providerID on Machines. Therefore although not necessarily this might definitely break providers as it's changing the equality contract. So we sure proceed taking that into consideration.

Also we should consistently update IndexKey().

**Comment by enxebre** [comment]
> @enxebre I'm not sure how IndexKey() is affected by this change

`// This is useful to use the providerID as a reliable index between nodes and machines
// as it guarantees the infra Providers contract.`


@jackfrancis indexKey was intentionally sticking to the equality contract which we are changing so it should be updated with it. Need to thing about impact on running clusters though, I guess it should be fine as the controller is restarted and will just reindex using the new criteria.

**Comment by enxebre** [comment]
>I think there's more background here that I'm not aware of, but from a naive point of view IndexKey() doesn't have any value if it simply returns the entire string (it will then be equivalent to the CloudProvider() method which just thinly wraps the cloudProvider string property)

indexKey is just an abstraction to be used to index (and fetch) machines/nodes by providerID in the controllers indexers. It must be unique and so It should use whatever we consider to be the equality contract, otherwise we could end up with one index mapping to multiple machines/nodes. If the impl of indexKey happens to be a thing wrapper that's fine.

**Comment by jackfrancis** [comment]
@enxebre got it, updated

**Comment by alexeldeib** [comment]
> CCM set on Nodes but capi providers set providerID on Machines. Therefore although not necessarily this might definitely break providers as it's changing the equality contract. So we should proceed taking that into consideration

Hm. That would require CCM and CAPI to disagree on what the providerID is for a given node, though? I suppose it could be something like -- CAPA sets provider ID using AZ, CCM doesn't, current logic evaluates those as the same node even though the strings are different? This is sort of why I wanted a sanity check on other providers. 

That would seems to require CAPI providers to be mutating the provider ID they get back from cloud providers (or perhaps CCM doing the same), which would seem like a world of hurt regardless of the contract. 

At a glance, seems like CAPA just reads whatever EC2 returns, which is roughly the same as Azure. Not sure how CCM for AWS generates provider IDs. https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/7a545cd1aa5d05e5a0364a49b3c700341737ca61/controllers/awsmachine_controller.go#L510-L511

Definitely want to make sure we get this lined up right either way.

**Comment by enxebre** [comment]
>Hm. That would require CCM and CAPI to disagree on what the providerID is for a given node, though? I suppose it could be something like -- CAPA sets provider ID using AZ, CCM doesn't, current logic evaluates those as the same node even though the strings are different? This is sort of why I wanted a sanity check on other providers.

@alexeldeib That's exactly right, so since we are changing the equality criteria this could technically break any provider out there. So at the very minimum this would require a note for providers implementers.

Also I agree with @killianmuldoon https://github.com/kubernetes-sigs/cluster-api/pull/6412#discussion_r849271035 we need to deprecate existing behaviour first as this is a public func.

**Comment by jackfrancis** [comment]
Correction to my comment above, the `IndexKey()` has now been made equivalent to `String()` (not `CloudProvider()`).

That change has been realized in the PR now, tests look good.

I think at this point we need to evaluate the potential for provider breakage, and if we can convince ourselves that this is actually fixing a capi bug based on the way that providerID URLs are set from their real-world authoritative sources, then we just want to make a communication plan so that any providers who need to update their code have time to do so.

File: controllers/noderefutil/providerid.go [hunk 2]
```diff
@@ -89,7 +88,7 @@ func (p *ProviderID) ID() string {
 
 // Equals returns true if both the CloudProvider and ID match.
```
Comment (CecileRobertMichon) [inline controllers/noderefutil/providerid.go:89]: comment should be updated to match new function (or we should make a new func as @killianmuldoon suggests)

Comment (jackfrancis) [reply controllers/noderefutil/providerid.go:89]: How might we deprecate this in favor of another func and not affect users? AFAICT this func is only used here:

https://github.com/kubernetes-sigs/cluster-api/blob/release-1.1/internal/controllers/machine/machine_controller_noderef.go#L196

If we created a new func and updated the reference in `(r *Reconciler) getNode` we would still be making a change for all providers, no?

Comment (killianmuldoon) [reply controllers/noderefutil/providerid.go]: I think we need to deprecate it and replace as it's a part of our public Go API, so we don't have control over who is using it and for what.

As for the behavioural change inside the CAPI method, I think the impact of that needs to be assessed separately. If there is an impact from changing the behaviour here that should be called out in the deprecation notice and/or on the migration guide.

Comment (CecileRobertMichon) [reply controllers/noderefutil/providerid.go]: @jackfrancis here's an example of deprecating a public func in CAPI https://github.com/kubernetes-sigs/cluster-api/pull/5545/files#diff-e25fb86695a8022fa1611100bd244f9b4185f28629d19830e83d18b4fa2ce710R41

And an example of adding to the migration guide for providers https://github.com/kubernetes-sigs/cluster-api/commit/f73a27729acf3bc2cce2479220e1460eb6f73efb#diff-531b360cba021aa4e0cf2df8f2ce8ead9601db1601bf516cab3c916a83622436R85

**Comment by vincepri** [comment]
/hold

I'm quite sure we had this behavior back in the v1alpha1 days and we had to change for something more structured given that some Cluster API providers can only set (guess?) a partial provider ID. Looking through the Cluster API AWS provider code, we're still doing that today https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/760e4e7651f46c29bf309607a0b6307570228988/pkg/cloud/scope/machine.go#L144-L148

Unless we have a contract requirement for every cloud provider to match their equivalent CCM/CPI codebase, we should keep this behavior and introduce a new one over time.

I might have missed this while going through the above messages, what is the problem we're trying to solve with this PR?

**Comment by alexeldeib** [comment]
> what is the problem we're trying to solve with this PR

The contract we rely on for provider ID comparison in CAPI is broken for Azure scale sets. The problematic logic is the logic in CAPI itself, not anything Azure does -- we set provider ID in line with CCM, and there would be no other reasonable way to differentiate these machines like using AZ in AWS.

See https://github.com/kubernetes-sigs/cluster-api/issues/4526

> In the existing providerID logic assumptions, cloudProvider and ID is what is used to compare, skipping the other segments of the providerID. See https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/noderefutil/providerid.go#L86

> **This makes the assumption - not necessarily true - that IDs in different regions/zones won't be reused by the cloud provider.**

**Comment by vincepri** [comment]
The part about the ID in different AZ being different seems fair to me, it has been a known issue for a while. Do we have any valid alternative of comparison we can make for these IDs? For example, if we were to start to take into account the AvailabilityZone/FailureDomain/Region, would that suffice?

**Comment by alexeldeib** [comment]
for some clarity -- with the current code the provider ID check in CAPI returns true when two VMs have IDs like:

/subscriptions/xxx/resourceGroups/xxx/Microsoft.Compute/virtualMachineScaleSets/**xxx**/1
/subscriptions/xxx/resourceGroups/xxx/Microsoft.Compute/virtualMachineScaleSets/**yyy**/1

even though these are different machines. 

I guess you could try doing both checks, but if you ever fail the stricter one (e.g. node doesn't exist yet), you can still pass the "less strict" one (using the example above, if yyy/1 didn't exist yet but that's the machine we were looking for).

**Comment by vincepri** [comment]
Let's enqueue this discussion topic for this week's community meeting as well, and make sure to reach out to cloud providers and mailing list to align providerID generation.

**Comment by alexeldeib** [comment]
> potentially alert on discrepancies

I particularly like this. Might be tricky with a webhook given the cross-object (cross-cluster? mgmt <-> workload) objects, but perhaps a condition + some sort of quick validation with kubectl would be nice. 

Will try to avoid suggesting much else in the weeds for now, but I think that could provide a nice path for e.g. new behavior opt-in -> you have a way to ensure you won't break -> new behavior by default (or something).

**Comment by CecileRobertMichon** [comment]
@jackfrancis do you have the bandwidth to take the steps outlined by Vince above?

**Comment by jackfrancis** [comment]
@vincepri @alexeldeib can we re-clarify what the "fallback behavior" is exactly? Reading through the above thread, it seems pretty clear that the current state of the changeset in this PR does not include any fallback.

**Comment by vincepri** [comment]
@jackfrancis A fallback could be defined as @alexeldeib pointed out above:

> I guess you could try doing both checks, but if you ever fail the stricter one (e.g. node doesn't exist yet), you can still pass the "less strict" one (using the example above, if yyy/1 didn't exist yet but that's the machine we were looking for).

That said, I think given that this is a breaking change and there could be undefined behaviors by running both tests, we should probably just run a big warning in the release notes and make sure providers that are affected update their code appropriately.

**Comment by sbueringer** [comment]
@vincepri Are you sure this will be part of v1.2? Feature freeze was planned to be on Monday (and the milestone applier usually assigns the milestone of the base branch on PRs, which is v1.3 at this point)

**Comment by jackfrancis** [comment]
+1 to pushing this out to 1.3

**Comment by CecileRobertMichon** [comment]
Hi all, we have CAPZ AKS users hitting this bug in the wild (see https://kubernetes.slack.com/archives/CEX9HENG7/p1658498393187099), any chances of getting this pulled in sooner rather than later?

**Comment by sbueringer** [comment]
> Hi all, we have CAPZ AKS users hitting this bug in the wild (see https://kubernetes.slack.com/archives/CEX9HENG7/p1658498393187099), any chances of getting this pulled in sooner rather than later?

If it is a breaking change (and continues to be one), I don't think we can cherry-pick it into a v1.2 patch release.

**Comment by CecileRobertMichon** [comment]
> If it is a breaking change (and continues to be one), I don't think we can cherry-pick it into a v1.2 patch release.

@sbueringer agreed, this is not a candidate for cherry-pick, but it would be good to see it get merged in main so it can be part of the next v1.3 minor release.

**Comment by vincepri** [comment]
What's the status of this PR?

**Comment by jackfrancis** [comment]
@vincepri I think we are determining whether @killianmuldoon's remarks about deprecation are required for queueing this up for a 1.3 release. If we feel this is an appropriate change for a minor release, it can be considered for final review.

**Comment by vincepri** [comment]
We should merge this PR and release it in v1.3 as a breaking code-behavioral change. It solves a major problem where two machines could result in the same ID, which could potentially break a lot of assumptions in the overall system.

/lgtm

**Comment by CecileRobertMichon** [comment]
/cc @Ankitasw @sedefsavas @richardcase 

To make sure CAPA is all set to accept this change 

/lgtm

**Comment by vincepri** [comment]
@fabriziopandini @sbueringer over to you for approval

**Comment by sbueringer** [comment]
> @fabriziopandini @sbueringer over to you for approval

I'll take a look early next week after the public holiday on Monday

**Comment by CecileRobertMichon** [comment]
@dthorsen the specific bug you're referring to should already be fixed in #6971

**Comment by sbueringer** [comment]
As far as I understand the situtation the PR sounds fine to me.

@jackfrancis  Let's please add a note to `v1.2-to-v1.3.md` to increase the chances infra providers will notice the change when they start adopting v1.3 alpha|beta|RC|GA

File: controllers/noderefutil/providerid_test.go [hunk 3]
```diff
@@ -115,19 +116,34 @@ func TestInvalidProviderID(t *testing.T) {
...
+	g.Expect(parsedAzure1.String()).To(Equal(inputAzure1))
+	g.Expect(parsedAzure1.ID()).To(Equal("default-template-control-plane-fhrvh"))
+	g.Expect(parsedAzure1.CloudProvider()).To(Equal(azure))
+
+	inputAzure2 := "azure:///subscriptions/4920076a-ba9f-11ec-8422-0242ac120002/resourceGroups/default-template/providers/Microsoft.Compute/virtualMachines/default-template-control-plane-fhrvh"
+	parsedAzure2, err := NewProviderID(inputAzure2)
+	g.Expect(err).NotTo(HaveOccurred())
+
+	// Test for equalitey
+	g.Expect(parsedAzure1.Equals(parsedAzure2)).To(BeTrue())
```
[the remark this answers is not in the document]
Comment (jackfrancis) [reply controllers/noderefutil/providerid_test.go:148]: Yes! done

**Comment by jackfrancis** [comment]
@sbueringer added a note to the 1.2-1.3 doc

PTAL and I think we're finally ready here

**Comment by vincepri** [comment]
Changes LGTM, @jackfrancis can you rebase?

**Comment by jackfrancis** [comment]
@vincepri done

**Comment by sbueringer** [comment]
Thank you very much!

/lgtm

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**